### PR TITLE
UrlBar and suggestion improvements

### LINKED ISSUE #4663: URL bar sometimes shows a part of domain
**Describe the issue you encountered:** URL bar sometimes shows a part of domain in the history after the displayed URLs. I am not sure how to reproduce the issue.

urlbar

**Expected behavior:** 

- Platform (Win7, 8, 10? macOS? Linux distro?): Windows 10

- Brave Version: 0.12.4 while I think I've seen this since several versions

- Any related issues:

Setting a milestone for now.

### LINKED ISSUE #4941: Selecting and Cutting the UrlBar text reverts the text back
**Did you search for similar issues before submitting this one?**

No

**Describe the issue you encountered:**

Selecting and cutting the whole url in the url bar restores it back.

**Expected behavior:**

The url bar should remain empty

- Platform (Win7, 8, 10? macOS? Linux distro?): Linux

- Brave Version: https://github.com/brave/browser-laptop/commit/497ecf8679e0bc14ae7d1e05e38f3b5a00733071

- Steps to reproduce:
    1. Open a website
    2. Hit CTRL+L -> CTRL+A -> CTRL+X
    3. Observe the text you have just cut restores back.

- Screenshot if needed:

- Any related issues:
https://github.com/brave/browser-laptop/issues/4219

### LINKED ISSUE #5025: CTRL+L wrong behaviour
**Did you search for similar issues before submitting this one?**

No.

**Describe the issue you encountered:**

Hitting CTRL+L (focus url bar) selects the current url bar text (which is correct) but also renders url bar suggestions (which is wrong) thus leading to wrong url in the current window, since the first autocomplete suggestion is highlighted.

For example in Chrome if you hit CTRL+L it will correctly only highlight the text and will render suggestions only if you start typing.

**Expected behavior:**

The Brave behaviour should match Chrome's

- Platform (Win7, 8, 10? macOS? Linux distro?): All

- Brave Version: 0.12.6-Preview 2

- Steps to reproduce:
    1. Open a website
    2. CTRL+L
    3. See suggestions come in

- Screenshot if needed:

http://webm.land/media/alKl.webm

- Any related issues:

### LINKED ISSUE #5043: Online suggestions in autocomple must be async
**Did you search for similar issues before submitting this one?**

No

**Describe the issue you encountered:**

Currently online suggestions that appear when you type in the URL bar are not async, blocking the suggestions list from rendering if you have slow internet connection. See the video http://webm.land/media/H1Yb.webm

The request for retrieving suggestions is async xhr, but the logic that renders the suggestions is not (which opens another bug that the first online suggestions that arrive are not displayed) ...

**Expected behavior:**

URL bar autocomplete must provide best experience for the user, instantly showing history and bookamrk matches and online suggestions when they arrive, just like Chrome and FF do.

- Platform (Win7, 8, 10? macOS? Linux distro?): All

- Brave Version: 0.12.6

- Steps to reproduce:
    1. Connect to slow network
    2. Start typing in the url bar
    3. Wait 5 min for suggestions

- Screenshot if needed:

- Any related issues:

### PR DESCRIPTION
**Pull request by Sh1d0w**
- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed).

Fixes #5025
FIxes #4941 
Fixes #5043 
Fixes #4663

cc @bbondy @diracdeltas

### CODE REVIEW

File: js/components/urlBar.js [hunk 1]
```diff
@@ -204,8 +197,13 @@ class UrlBar extends ImmutableComponent {
       case KeyCodes.BACKSPACE:
         this.hideAutoComplete()
         break
+      // Do not trigger rendering of suggestions if you are pressing alt or shift
+      case KeyCodes.ALT:
+      case KeyCodes.SHIFT:
```
Comment (diracdeltas) [inline js/components/urlBar.js]: shouldn't this also include most other modifier keys?

File: test/components/urlBarTest.js [hunk 2]
```diff
@@ -88,20 +87,5 @@ describe('urlBar', function () {
             .then((val) => val === 'aboutx')
         })
     })
-
-    it('it does not change input after focus until keydown', function * () {
```
Comment (diracdeltas) [inline test/components/urlBarTest.js]: please change this test to test that focus doesn't bring up any suggestions, thanks

File: js/components/urlBar.js [hunk 3]
```diff
@@ -281,6 +275,9 @@ class UrlBar extends ImmutableComponent {
     this.clearSearchEngine()
     this.detectSearchEngine(e.target.value)
     this.keyPressed = false
+    if (e.target.value.length === 0) {
```
Comment (diracdeltas) [inline js/components/urlBar.js]: just to be paranoid-safe, check that `e.target.value` exists before calling `e.target.value.length`

File: js/components/urlBar.js [hunk 4]
```diff
@@ -38,16 +38,14 @@ class UrlBar extends ImmutableComponent {
     this.onFocus = this.onFocus.bind(this)
     this.onBlur = this.onBlur.bind(this)
     this.onKeyDown = this.onKeyDown.bind(this)
-    this.onCut = this.onCut.bind(this)
-    this.onKeyUp = this.onKeyUp.bind(this)
+    this.onChange = this.onChange.bind(this)
```
Comment (diracdeltas) [inline js/components/urlBar.js]: this is minor, but the `s/onChange/onKeyUp` change should be kept because it helps readability

**Comment by diracdeltas** [comment]
this seems to have introduced a bug where the URL bar suggestions appear and flicker when the tab is switched. str:

1. open tab, go to bing.com
2. open tab, go to twitter.com
3. switch back to tab 1, autosuggest renders automatically
4. switch to tab 2, autosuggest renders there as well

**Comment by Sh1d0w** [comment]
@diracdeltas Yes I know I am preparing commit atm. Will let you know when it is ready for review. Thanks

File: js/components/urlBar.js (hunk 1 above)
Comment (Sh1d0w) [inline js/components/urlBar.js]: I have added two more. If you think I missed some, please let me know I will add :)

**Comment by Sh1d0w** [comment]
@diracdeltas All fixed now, added tests as well :)

**Comment by Sh1d0w** [comment]
Updated the PR with the fix for https://github.com/brave/browser-laptop/issues/5043 and closed the other PR so all is in one place.

File: test/components/urlBarTest.js [hunk 5]
```diff
@@ -89,18 +88,11 @@ describe('urlBar', function () {
...
 
-    it('it does not change input after focus until keydown', function * () {
-      yield this.app.client
-        .keys('https://brave.com/ ')
-        .keys('\uE007')
-        .waitForElementFocus(activeWebview)
-        .ipcSend('shortcut-focus-url')
-        .getValue(urlInput).then((val) => assert(val === 'https://brave.com/'))
-        .keys('\uE015')
+    it('does not show suggestions on focus', function * () {
```
Comment (diracdeltas) [inline test/components/urlBarTest.js]: i think you want to type some letters before focusing, otherwise it wouldn't show suggestions anyway

Comment (Sh1d0w) [inline test/components/urlBarTest.js]: Ah, I am so sleepy. Fixed, thanks

File: test/components/urlBarTest.js [hunk 6]
```diff
@@ -89,7 +89,9 @@ describe('urlBar', function () {
     })
 
     it('does not show suggestions on focus', function * () {
-      yield this.app.client.ipcSend('shortcut-focus-url')
+      yield this.app.client
+        .keys('brave')
```
Comment (diracdeltas) [inline test/components/urlBarTest.js]: just to be sure this actually brings up suggestions, please add
```
.waitUntil(function () {
            return this.isExisting(urlBarSuggestions).then((exists) => exists === true)
```
here

**Comment by diracdeltas** [comment]
otherwise lgtm once this is rebased, assuming travis mostly passes. thanks!

File: test/components/urlBarTest.js (hunk 6 above)
Comment (Sh1d0w) [inline test/components/urlBarTest.js]: Done

**Comment by Sh1d0w** [comment]
@yan rebased, no worries.

THe Url bar feels so fast and slick now, I've built a custom package for me with it so I can test it :)

**Comment by diracdeltas** [comment]
@Sh1d0w i think there are some travis test failures related to this change. specifically 14-17 in https://travis-ci.org/brave/browser-laptop/builds/169690847

**Comment by Sh1d0w** [comment]
@diracdeltas Travis constanlty fails different tests. See for example here https://travis-ci.org/brave/browser-laptop/builds/169702712 which is from this PR (not related to any navbar changes) https://github.com/brave/browser-laptop/pull/5056

If I execute the tests locally everything passes, also if I manually test the failed tests (for example typing the first char) it works just ok.

**Comment by diracdeltas** [comment]
@Sh1d0w yeah, travis is not reliable yet so i mostly look at tests which seem like they may be related. in this case `npm run test -- --grep='urlbarSuggestions'` and `npm run test -- --grep='navigationBar'` do not pass locally for me. could plausibly be unrelated to this PR, in which case i'll fix it later.

**Comment by diracdeltas** [comment]
actually all those failures from this PR exist in https://travis-ci.org/brave/browser-laptop/builds/169696262 as well except for:
```
  14) navigationBar typing "before all" hook: _callee46:
     Error: element (.urlBarSuggestions li) still not existing after 5000ms
```

so i think that one is probably related, but the others aren't.

**Comment by Sh1d0w** [comment]
@diracdeltas I will look into this and will get back to you with my findings. Thanks

**Comment by Sh1d0w** [comment]
@diracdeltas The tests are fixed, it should pass on your local now. Let's see what is travis thinking about this :)

**Comment by Sh1d0w** [comment]
@diracdeltas Actually the test for typing a single character is failing and thats why `14)` is failing as well. Idk what's the case there since if you run locally my branch you can see taht if you type a single character in the URL bar the suggestions appear. 

I guess it is timeout issue since it only fails in the test but I could use some help here if you don't mind :)

**Comment by Sh1d0w** [comment]
@diracdeltas The bug is fixed all tests should pass now. Let me know if you find any other issue.

**Comment by Sh1d0w** [comment]
@diracdeltas @bbondy Because of the changes made in https://github.com/brave/browser-laptop/issues/3049 most of the test were failing because now the order of history entries have changed.

I fixed those tests in my PR, but please make sure in future when merging stuff that tests pass, because as you can see then we have trouble in the following PR's

Thaks guys.

**Comment by diracdeltas** [comment]
LGTM, thanks!

**Pull request merged**